aten/src/ATen/native/sparse/SparseTensorMath.cpp

```diff
@@ -799,4 +800,241 @@ Tensor sspaddmm(const Tensor& self, const Tensor& mat1, const Tensor& mat2,
   return result;
 }
 
+// --------------------------------------------------------------------
+// sparse.sum()
+// --------------------------------------------------------------------
+Tensor _sparse_sum(const SparseTensor& input) {
+  return input.values().sum();
+}
+
+Tensor _sparse_sum(const SparseTensor& input, ScalarType dtype) {
+  return at::_sparse_sum(input.to(dtype));
+}
+
+Tensor _sparse_sum(const SparseTensor& input, IntList dims_to_sum) {
+  return at::_sparse_sum(input, dims_to_sum, false);
+}
+
+Tensor _sparse_sum(const SparseTensor& input, IntList dims_to_sum, ScalarType dtype) {
+  return at::_sparse_sum(input.to(dtype), dims_to_sum, false);
+}
+
+Tensor _sparse_sum(const SparseTensor& input, IntList dims_to_sum, bool keepdim) {
+
+  AT_CHECK(input.is_coalesced(), "To support autograd, input SparseTensor has to be coalesced.");
```

Comment:
I think it's totally reasonable to not implement the non-coalesced case. However, I don't think the justification given here is quite right. It would totally be possible, in principle, to support sparse sum on non-coalesced tensors: the non-dims-to-sum case is able to do precisely this! (And in principle, coalescing the tensor and then calling this function would get the correct autograd formula, right?) So this is more of a, "We didn't implement it situation" than "Autograd is not meaningful when the input is not coalesced."

Reply:
Why not calling `input_ = input.coalesce()` here? Similar to how we make things contiguous before calling many kernels?

Reply:
@ezyang Yeah, totally agreed! Sorry about the wordy. And yes, we could support non-coalesced case, but just not there yet. I will change the message. 

@SsnL You mean to mutate the input tensor? Is it legitimate to do that? I thought for the case of  `contiguous` it is because backwards can handle for non-contiguous inputs.

Reply:
ok, I see, I will keep the `is_coalesced()` check for input then